test/cluster/basic_test.go

```diff
@@ -101,3 +101,348 @@ func TestBasicCluster(t *testing.T) {
+			require.Contains(t, err.Error(), "Redirect blocked in test client: url:")
+			require.Contains(t, err.Error(), fmt.Sprintf("referrer: '%s/data/tx'", srv.URL()))
+			txRedirected++
+			followerServer = srv
+		} else {
+			t.Logf("tx submitted: %s, %+v", txID, rcpt)
+			txOK++
+			leaderServer = srv
+			leaderValue = i
+			t.Logf("Leader is: %s", srv.URL())
+		}
+	}
+	require.Equal(t, 1, txOK)
+	require.Equal(t, 2, txRedirected)
+
+	require.NoError(t, c.ShutdownServer(followerServer))
+
+	txID, rcpt, err := leaderServer.WriteDataTx(t, worldstate.DefaultDBName, "bob", []byte{uint8(leaderValue), uint8(leaderValue)})
+	if err != nil {
+		t.Logf("txID: %s, error: %s", txID, err)
+	} else {
+		t.Logf("tx submitted: %s, %+v", txID, rcpt)
+	}
+
+	require.NoError(t, c.StartServer(followerServer))
+
+	time.Sleep(10 * time.Second)
+
+	dataEnv, err := followerServer.QueryData(t, worldstate.DefaultDBName, "bob")
+	require.NoError(t, err)
+	dataResp := dataEnv.GetResponse()
+	t.Logf("data: %+v", dataResp)
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - wait for one to be the leader.
+// - submit a tx to each of the 3, with a client that catches redirects.
+// - stop the leader.
+// - wait for new server to be the leader.
+// - submit a tx to the two remaining servers.
+// - start the stopped server.
+// - make sure the stopped server is in sync with the transactions made while it was stopped.
+func TestNodeRecoveryLeader(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	time.Sleep(10 * time.Second) //TODO replace by a call to detect a leader
+
+	var followerValue int
+	var followerServer *setup.Server
+	var leaderServer *setup.Server
+	var txOK int
+	var txRedirected int
+	for i, srv := range c.Servers {
+		txID, rcpt, err := srv.WriteDataTx(t, worldstate.DefaultDBName, "alice", []byte{uint8(i), uint8(i)})
+		if err != nil {
+			t.Logf("txID: %s, error: %s", txID, err)
+			require.Contains(t, err.Error(), "Redirect blocked in test client: url:")
+			require.Contains(t, err.Error(), fmt.Sprintf("referrer: '%s/data/tx'", srv.URL()))
+			txRedirected++
+			followerServer = srv
+			followerValue = i
+		} else {
+			t.Logf("tx submitted: %s, %+v", txID, rcpt)
+			txOK++
+			leaderServer = srv
+			t.Logf("Leader is: %s", srv.URL())
+		}
+	}
+	require.Equal(t, 1, txOK)
+	require.Equal(t, 2, txRedirected)
```

Comment:
This part repeats itself in many tests, and is basically "submit to detect a leader". It can be done across all servers, or possibly a subset when some servers are knows to be down. You can put it in a function.
If there is no leader at all; for instance it takes some time to elect, then the function will fail, but you can then retry to submit-to-detect-a-leader until it eventually succeeds... for example, retry every 1s for 30s total. This can be useful when you do the rolling restarts tests. 

You can also use `assert.Eventually(t, isLeaderCond, 30*time.Second, 1*time.Second)`  to automate this condition, see e.g. https://github.com/hyperledger-labs/orion-server/blob/a5db614f1044993886df4719e3ac14d9a65aa5ad/internal/replication/blockreplicator_3node_test.go#L64